### DISCUSSION TITLE
Creating new wordnets / wordnets extensions

### DISCUSSION THREAD

**Discussion by francis-dion** [goodmami/wn#165] [Q&A] [answered]
I'm looking to create brand new wordnets and/or wordnet extensions, specifically along those two scenarios:
1. Incremental, community-driven building of wordnets for aboriginal languages.
2. Industry or organization-specific wordnets or wordnet extensions (?) to capture semantic relationships, definitions and examples specific to a specific context/culture/social group.

Both scenarios would likely produce very sparse space wordnet instances heavyly reliant on "main stream" wordnets for broader application/context.

Simple, high-level set of capabilities I'm looking for include:
- Initiating a new (empty) wordnet.
- Adding, deleting and editing new content (wn-editor?)
- Importing and merging part or all content from another wordnet into one already loaded.
- Comparing two wordnets for gaps, overlaps, etc.

Can wn or wn-editor be used for that purpose? Should they? 
Are there other tools/library/services better suited for these purposes? 

Any thoughts, ideas and suggestions (including of better forums/communities where to post this query :-) will be much welcomed!

**Comment by goodmami** [comment, marked as answer]
For now, Wn can be used to facilitate the last two bullet points (partial import, comparing wordnets) with some scripting as they are not out-of-the-box features. Wn does not currently help with wordnet creation/modification, but the other project you mentioned, [wn-editor](https://github.com/Hypercookie/wn-editor) aims to do just that.

Regarding the sparse wordnets reliant on more complete ones, this is in fact a primary purpose of Wn. Most non-English wordnets were created using the "expand" methodology where non-English lemmas are placed on Princeton WordNet (PWN) synsets. While this causes the new wordnets to adapt to the conceptual structure of English (as defined by PWN), it helped the wordnets develop quickly. Wn can handle these "expand" wordnets with or without the base one (although it helps to have the base, which is invariably but not necessarily PWN 3.0, for synset relation traversal). Secondly, Wn accommodates "extension" lexicons which are meant to augment some base language to, for instance, add domain-specific jargon or relationships. Unfortunately there are not yet many examples of this (there's [taboown](https://github.com/bond-lab/taboown), but it is not yet encoded as a [WN-LMF 1.1](https://github.com/globalwordnet/schemas) extension).

**Reply by francis-dion** [reply]
Thanks for the quick response.
My third bullet was imprecise as I meant "import and merge", not just "import". I'm guessing I can import with Wn and merge with Wn-editor. I will try that shortly.

I'm curious about what you mean by "Wn accommodates extension lexicons". Also not sure what would be required to encode taboown a a WN-LMF 1.1 extension. (as you see, very newby here :-).

More broadly, for my purpose, I wonder if I'm better off (a) adding new lexicons/content directly in Wn database with Wn-editor, or (b) "bite the bullet" and create new lexicons directly (following guidelines from here https://globalwordnet.github.io/schemas/) and then import them in Wn.

If going with (a) above, I'll then sooner than later be interested in exporting these lexicons in XML for sharing etc. Any plans/guidelines to that effect?

**Reply by goodmami** [reply]
> I'm curious about what you mean by "Wn accommodates extension lexicons". Also not sure what would be required to encode taboown a a WN-LMF 1.1 extension.

There is a description in [McCrae et al., 2021](https://aclanthology.org/2021.gwc-1.11.pdf), Section 3.2.2. Basically the `<Lexicon>` XML element needs to be `<LexiconExtension>` and it needs to specify which base lexicon it is extending. When things like added relations refer to things in the base lexicon, surrogate ("external") elements need to be created in the extension so the references can be resolved.

> More broadly, for my purpose, I wonder if I'm better off (a) adding new lexicons/content directly in Wn database with Wn-editor, or (b) "bite the bullet" and create new lexicons directly (following guidelines from here https://globalwordnet.github.io/schemas/) and then import them in Wn.

Wn-editor is very new and I haven't had a chance to try it yet, so I cannot really recommend on way or the other. I would note, however, that Wn currently loads extension lexicons but, unlike with regular lexicons, it cannot export them. So if you want to make an extension and build it inside Wn's database, you won't be able to distribute a WN-LMF xml file until that is fixed (see #103).

You could, however, use Wn's [wn.lmf](https://wn.readthedocs.io/en/latest/api/wn.lmf.html) module to create an intermediate object (the internal representation between XML and the database) which can be written to XML. The data structure are simple Python dictionaries, but the module has classes for type-checking them and these can also be used to create them:

```pycon
>>> import wn.lmf
>>> wn.lmf.Count(value=3)
{'value': 3}
>>> wn.lmf.Example(text="this is an example")
{'text': 'this is an example'}
```

These are not currently considered a public part of the module and they are therefore not documented, so please inspect [lmf.py from line 161](https://github.com/goodmami/wn/blob/main/wn/lmf.py#L161) and let me know if you have questions.

**Reply by francis-dion** [reply]
Thanks for the McCrea et al. reference, it is very helpful. This though is a concern for my purpose:

> Destructive extensions that, for instance, remove entries or senses from a lexicon or selective dependencies that exclude certain relations are left to future work.

Do you know of any plans on extending the concept of extensions to support "destructive" ones? Anyway, I'll run a few experiments and test some some ideas I have for workarounds.

I didn't realize the wn object had an "export" method. Just curious, what was the purpose of implementing this considering there isn't built-in functionality to update a lexicon?

**Reply by goodmami** [reply]
> Do you know of any plans on extending the concept of extensions to support "destructive" ones?

No plans currently. I'd imagine it would be implemented as some kind of block-list. I.e., the entries would remain in the database but would not turn up in queries when the extension is used.

> Just curious, what was the purpose of implementing this considering there isn't built-in functionality to update a lexicon?

There is/was an intention to support editing lexicons, but it had been backburnered because of, until recently, little demand for the feature and because it would add significant complexity to the codebase, thus increasing the maintenance burden. The export functionality also served as a useful check that we were modeling all information in a WN-LMF lexicon. E.g., if we could round-trip the full lexicon, we could be reasonably sure that no info was getting lost.
